Adding GW loss function

janetrbarclay opened the pull request:
This PR includes the current code for running the groundwater loss function (turned off / on with the loss_type parameter in the config file). 

The Snakefile has an additional preprocessing and 2 additional postprocessing rules related to the GW metrics, and there are additional parameters in the config file that are passed to the training rules (which were modified to accept the additional inputs). 

Currently the code for pre-/post- processing of temperature signals are in a separate script (gw_utils.py) but these could be moved into the existing pre-/post- processing scripts if that's better. The loss function code itself was added to loss_functions.py.

I'm using two packages that aren't currently part of the rgcn conda environment - seaborn and statsmodels. Neither is essential to the loss function itself, but are used in the preprocessing (statsmodels allows easy calculation of confidence intervals for regression coefficients) and postprocessing (seaborn).

I'd be happy to talk through the additions if that would be helpful.

- Janet

aappling-usgs commented on Snakefile_full:
```diff
@@ -0,0 +1,134 @@
+import os
```
What's the difference between Snakefile and Snakefile_full? Are both used?

janetrbarclay replied:
Thanks. It looks like I ended up with a bunch of extra files in the repo (the cached ones and some scratch ones like the ipynb files). I'll get those deleted.

janetrbarclay replied:
I think all the extra files are removed. Let me know if you find any that I missed.

aappling-usgs reviewed (commented):
Hi Janet, I'm excited to see your PR is up! I'll leave most of the reviewing to @jsadler2 or @jzwart , but a preliminary change you could make would be to remove the caching folders and folder contents from the PR. I see files from `.ipynb_checkpoints`, `.snakemake`, `river_dl/.ipynb_checkpoints`, and `river_dl/__pycache__` that aren't needed on GitHub. You can git remove them, commit the removal, and then add those folders to .gitignore.
  I also see a few .ipynb files - are these integral to your analysis (and specifically to the part of the analysis that should be going in this generalized model code repo)? In addition to wondering about their application, I'll point out that .ipynb files are difficult to view in a plaintext editor such as GitHub, so I'd say in general we prefer plain .py files when possible.

jzwart commented on river_dl/gw_utils.py:
```diff
@@ -0,0 +1,451 @@
+    #Phi = (3/2)* pi - atan (b/a) - in radians
+    
+    #convert the date to decimal data
+    date_decimal = [float(x)/365 for x in ((Date-np.datetime64('1980-10-01'))/np.timedelta64(1, 'D'))]
```
where does the `1980-10-01` date come from? Could this be supplied as an argument with 1980-10-01 as default?

janetrbarclay replied:
this is the reference date from which decimal days (1 year  = 1) are calculated. The actual date shouldn't matter, but if it were changed to a summer start then the phase calculation would be slightly different (the inverse trig functions have to be modified for being in a different quadrant of the circle) 

I think it's probably best for it to be not easily modifiable (so not making it an argument), but I'll add a comment to the code.

janetrbarclay replied:
I added a function that creates the decimal year (it was previously in 2 places) and then added comments and a citation re: the reference date and what would need to change if the time of year of the reference date were changed.

jzwart commented on river_dl/gw_utils.py:
```diff
@@ -0,0 +1,451 @@
+    date_decimal = [float(x)/365 for x in ((Date-np.datetime64('1980-10-01'))/np.timedelta64(1, 'D'))]
+    
+    #remove water temps below 1C or above 60C to avoid complex freeze-thaw dynamics near 0 C and because >60C is likely erroneous  
+    if isWater:
+        temp = [x if x >=1 and x<=60 else np.nan for x in temp]
```
could also be supplied as an argument, but seems ok hard coded. >60 seems extremely high, maybe make the cutoff lower? like 40 or 45?

janetrbarclay replied:
I lowered the cutoff to 45

jzwart commented on river_dl/gw_utils.py:
```diff
@@ -0,0 +1,451 @@
+    return amp, phi, amp_low, amp_high, phi_low, phi_high
+
+
+def annTempStats(thisData):
```
`ann` reminds me of Artificial Neural Network, consider changing to `annualTempStats`

jzwart replied:
Also maybe use `_` for function naming rather than camel case (e.g. `ann_temp_stats`). I think it's more consistent with the rest of your gw functions and more consistent with most of the other `river-dl` functions.

janetrbarclay replied:
I've change the function to annual_temp_stats

jzwart commented on river_dl/gw_utils.py:
```diff
@@ -0,0 +1,451 @@
+        
+        #get the water temp properties
+        #ensure sufficient data
+        if np.sum(np.isfinite(thisData['seg_tave_water'][:,i].values))>(365): #this requires at least 1 yr of data
+            waterDF = pd.DataFrame({'date':thisData['date'].values,'seg_tave_water':thisData['seg_tave_water'][:,i].values})
```
could supply the water temperature variable as an argument - `seg_tave_water` is PRMS-SNTemp specific and it would be good to be able to calculate these variables for other models  with different variable names

janetrbarclay replied:
I did this for the PRMS air temp and water temp, and for the observed water temp

jzwart commented on river_dl/gw_utils.py:
```diff
@@ -0,0 +1,451 @@
+    delPhi_low_obs = [(water_phi_low_obs[x]-air_phi_high[x])*365/(2*math.pi) for x in range(len(water_amp_obs))]
+    delPhi_high_obs = [(water_phi_high_obs[x]-air_phi_low[x])*365/(2*math.pi) for x in range(len(water_amp_obs))]
+    
+    #remove Ar >1.1
+    delPhi_obs=[delPhi_obs[i] if Ar_obs[i] <=1.1 else np.nan for i in range(len(delPhi_obs))]
+    Ar_obs = [x if x <= 1.1 else np.nan for x in Ar_obs]
```
Could you add some more description of what `Ar` is and why there is a cutoff of 1.1?

janetrbarclay replied:
I've added this to the file.

jzwart commented on river_dl/gw_utils.py:
```diff
@@ -0,0 +1,451 @@
+    Ar_obs = [x if x <= 1.1 else np.nan for x in Ar_obs]
+    
+    #remove delPhi <-10
+    Ar_obs = [Ar_obs[i] if delPhi_obs[i] >=-10 else np.nan for i in range(len(Ar_obs))]
+    delPhi_obs = [x if x >=-10 else np.nan for x in delPhi_obs]
+    
+    #reset delPhi -10 to 0
+    delPhi_obs = [x if x > 0 else 0 if np.isfinite(x) else np.nan for x in delPhi_obs]
```
Could you also add some more comments here about what these QA/QC portions are doing?

janetrbarclay replied:
I've added this to the file.

jzwart commented on river_dl/gw_utils.py:
```diff
@@ -0,0 +1,451 @@
+    water_amp_preds = []
+    water_phi_preds = []
+    for thisSeg in segList:
+        amp, phi, amp_low, amp_high, phi_low, phi_high = amp_phi(thisData.loc[thisData.seg_id_nat==thisSeg,"date"].values,thisData.loc[thisData.seg_id_nat==thisSeg,"seg_tave_water"],isWater=True)
```
again could make `seg_tave_water` an input variable

janetrbarclay replied:
thanks. I updated this one too.

jzwart reviewed (commented):
Cool work @janetrbarclay ! I've added a few comments. 

I also wonder if some of the changes to the `snakemake`  and `config` files should be committed to the repo.  I think it might make sense to have those project-specific files live in another repository (e.g. a gw-specifc repo) and the snakemake + config files that live in the river-dl repo should be more of a template that could be used for many projects. It is good to see the example of incorporating gw loss into the training though. 

There are also a few functions that may need to change if / when #104 is merged into the repo.

aappling-usgs commented:
Great thoughts, @jzwart . @janetrbarclay , let me know if you'd like any help working through the git conflicts noted above - sometimes they're straightforward, other times not so much.

janetrbarclay commented:
Thanks. I should have mentioned this, but I paused on working through the conflicts to incorporate the recent updates to river-dl and to get my presentation ready for this afternoon. It's on my list for after that and I'll reach out if I have questions.

Thanks!

Janet

------------------------------------------------
Janet Barclay
U.S. Geological Survey
New England Water Science Center
Connecticut Office
101 Pitkin St.
East Hartford, CT 06108

Phone (office) 860 291-6763
Fax 860 291-6799
Email ***@***.******@***.******@***.***>
https://www.usgs.gov/staff-profiles/janet-barclay

________________________________
From: Alison Appling ***@***.***>
Sent: Tuesday, June 15, 2021 11:10 AM
To: USGS-R/river-dl ***@***.***>
Cc: Barclay, Janet R ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [USGS-R/river-dl] Adding GW loss function (#105)




 This email has been received from outside of DOI - Use caution before clicking on links, opening attachments, or responding.



Great thoughts, @jzwart<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjzwart&data=04%7C01%7Cjbarclay%40usgs.gov%7C500be1dac6a749131cf908d9300fbe62%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637593666444769863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=6qNak8%2FQTXLmnqYuteKAv7Mmbvr24N3%2F6dg%2FbxTNQMc%3D&reserved=0> . @janetrbarclay<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjanetrbarclay&data=04%7C01%7Cjbarclay%40usgs.gov%7C500be1dac6a749131cf908d9300fbe62%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637593666444779818%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=LqOpdHdMp%2Bp0nYO9SgcybqO8MzMgzZiAL2peoqPg6ZE%3D&reserved=0> , let me know if you'd like any help working through the git conflicts noted above - sometimes they're straightforward, other times not so much.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FUSGS-R%2Friver-dl%2Fpull%2F105%23issuecomment-861583105&data=04%7C01%7Cjbarclay%40usgs.gov%7C500be1dac6a749131cf908d9300fbe62%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637593666444779818%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=ejosSX4LE3muzDGWFJxYFqiSGKJb5THTAADuNslGN3M%3D&reserved=0>, or unsubscribe<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAA5H7UC4WCKHWALC4OWL3BLTS5UNXANCNFSM4562N5RQ&data=04%7C01%7Cjbarclay%40usgs.gov%7C500be1dac6a749131cf908d9300fbe62%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637593666444789777%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=TU6pKsltG9UHEYY%2BEiHxmGqpU8E6RMSKmAfVJxE96Rk%3D&reserved=0>.

janetrbarclay commented on .gitignore:
```diff
@@ -1,3 +1,12 @@
 data/
 *.npz
+DRB_data/
+DRB_output*/
```
done

janetrbarclay commented on Snakefile:
```diff
@@ -1,137 +1,197 @@
+                  test_start_date=config['test_start_date'],
+                  test_end_date=config['test_end_date'], 
+                  gwVarList = config['gw_vars'],
+                  out_file=output[0],
+                  out_file2=output[1])
```
I've combined them into 1 output file

janetrbarclay commented:
@jzwart Could you take a look at this again? I think I've addressed most of your comments. One that I didn't touch was keeping the Snakefile and config files out of the repo. That makes sense. How do I remove those files from this PR?

jzwart commented:
@janetrbarclay  sorry I'm just getting to this now - I was out on vacation and then buried in project work for a forecasting project. I'll look through all the new changes

jzwart commented on river_dl/train.py:
```diff
@@ -28,8 +29,11 @@ def train_model(
     loss_func,
     out_dir,
     model_type="rgcn",
+    loss_type="GW",
```
what would be the other types of input if not `GW` ?

janetrbarclay replied:
at the time I wrote this, anything other than the "gw" (case insensitive) defaulted to rmse. (now it would go to whatever is specified in the config file.) A couple possible options:

1. keep it as is and ensure there is adequate documentation in the code
2. switch it to a GW flag (yes / no on whether to use the GW loss function) instead of having the variable name suggesting that other possible options exits
3. incorporate the GW loss function as an option in the loss_func parameter. This would require modifying the code slightly to allow specifying different loss functions for pre-training and fine-tuning. 

Option 3 keeps things the cleanest and allows for other loss function configurations in the future, so that may be the way to go. But maybe that should be a new PR after finishing up this one?

jzwart replied:
Yeah I agree that option 3 seems like the best option. Would be good to open up a new PR for that functionality too 👍

janetrbarclay replied:
So we'll leave this as is for this PR, yes?

jzwart replied:
yes

jsadler2 replied:
Is there a reason you aren't defining the loss function in the Snakefile and then passing the function as a parameter to the `train` function?

janetrbarclay replied:
yes - that functionality was added to the Snakefile after this PR was started. I was thinking of switching to that with a new PR after finishing up this one. That will also involve adding an option to specify different loss functions for the pre-training and fine-tuning b/c the GW loss is only used for the fine-tuning.

janetrbarclay replied:
This comment seemed to be missing on my view from the PR thread so I'm reposting it:
yes - that functionality was added to the Snakefile after this PR was started. I was thinking of switching to that with a new PR after finishing up this one. That will also involve adding an option to specify different loss functions for the pre-training and fine-tuning b/c the GW loss is only used for the fine-tuning.

jsadler2 replied:
sounds good to me. for the next PR, I think it'd be good to include the option to pass in separate loss functions for pretraining and fine tuning. If possible, it'd be nice if we could make it so if someone just passes in `loss_func` it's used for both phases but if they have a finetune-specific or pretrain-specific loss function it uses that.

jzwart commented:
> One that I didn't touch was keeping the Snakefile and config files out of the repo. That makes sense. How do I remove those files from this PR?

To be clear, I think there should still be some sort of Snakefile and config file in the `river-dl` repo, I just don't think it should be project-specific. If someone using `river-dl` did want to see how to construct a Snakefile with gw loss functions, it might be nice to have an example in `river-dl`, but I'm not sure it should be the default. So I think there might be a few options: 

1. Remove your gw-specific Snakefile and config file from your local repo, add back the original Snakefile & config file to your local repo, commit & push (that should remove the modified Snakefile and config file from this PR)
2. Comment out the lines that are specific to gw losses in the Snakefile / config file and try to get as close as possible to the original default Snakefile / config file 
3. Save the gw-specific Snakefile and config file as `Snakefile_gw` or something similar, and add back in the original Snakefile to your local repo. This way, the original Snakefile would still be in river-dl and people will have an example of what including gw losses would look like. 

I think I'd prefer option 3, and then 1, and then 2. But it'd be good to hear what others think

janetrbarclay commented:
It would be pretty easy to do option 3 and that makes sense to me. Having alternative snakefiles and config files could be a good strategy in general for demonstrating use of the less routine (or project-specific) features of river-dl. (Could be a way to demonstrate use of the command line training rule as well.)

janetrbarclay commented:
@jzwart  I've added back the existing Snakefile and config.yml and tagged the gw ones at Snakefile_gw and config_gw.yml. Let me know if there's anything else I missed for this PR.

jzwart commented:
Thanks @janetrbarclay , I think it looks good to me!

janetrbarclay commented:
@jzwart So is it good to merge? If so, how do we do that? I don't see a link for it so either I'm missing something or I don't have access to do it.

jzwart commented:
Yeah I think it's good to merge. @jsadler2 , is @janetrbarclay a collaborator on the repo and has merge privileges? I can merge otherwise

jzwart commented:
pinging @jsadler2 on this again. You think it's alright to merge? and also curious why Janet can't merge

jsadler2 commented on config.yml:
```diff
@@ -1,31 +1,42 @@
-pt_epochs: 1
-ft_epochs: 1
-hidden_size: 20
+# Input files
```
Should this file be changed *and* there be a `config_gw.yml` file? maybe I'm missing it, but they look the same to me.

janetrbarclay replied:
I originally had a modified config.yml. Jake suggested keeping it as it had been and making a project specific one, so I saved my config.yml to config_gw.yml and added the original config.yml back in. The config_gw.yml file has a couple extra lines (gw_vars, lamb2, lamb3, loss_type)

jsadler2 reviewed (commented):
Hi @janetrbarclay - sorry for being so slow on checking this out. I have a few comments above that I think we should resolve before merging.

janetrbarclay commented on Snakefile_sens:
```diff
@@ -0,0 +1,205 @@
+import os
```
This was a sensitivity analysis I started. I've moved it to the project specific repo

janetrbarclay commented on Snakefile_sens:
```diff
@@ -0,0 +1,205 @@
+        ),
+        
+
+rule prep_io_data:
```
that is a cool feature. I ended up using the module option instead b/c there were options to not use all the rules (or to modify them). As is currently written, I slightly modify the train model rules to pass in lambda2 and lambda3 so I needed to be able to not use that rule from the original snakefile. For now I just set the outputs to "" so they aren't in any workflow. Once the gw loss function is passed as a function to the train rule, then it would probably be just a modified input value and there wouldn't be a gw-specific training rule

janetrbarclay commented on Snakefile_gw:
```diff
@@ -0,0 +1,199 @@
+import os
```
sure. I added a note that this is an example of how to use the GW loss function

janetrbarclay commented on Snakefile_gw:
```diff
@@ -0,0 +1,199 @@
+        
+
+
+rule prep_io_data:
```
that is a cool feature. I ended up using the module option instead b/c there were options to not use all the rules (or to modify them). As is currently written, I slightly modify the train model rules to pass in lambda2 and lambda3 so I needed to be able to not use that rule from the original snakefile. For now I just set the outputs to "" so they aren't in any workflow. Once the gw loss function is passed as a function to the train rule, then it would probably be just a modified input value and there wouldn't be a gw-specific training rule

janetrbarclay commented:
@jsadler2  I think I responded to all your comments. Let me know if there is anything I missed. My response about why the loss function isn't defined in the Snakefile isn't showing up on the Conversation tab, but it does show up on the "Files Changed" tab.

jsadler2 commented:
@janetrbarclay - thanks for making those changes. I think it's good to merge, now. I will do that.
